File: Week3/ArrayProblem/FourSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class FourSum {
/*
 * Given an array nums of n integers, return an array of all the unique quadruplets [nums[a], nums[b], nums[c], nums[d]] 
 * such that: 0 <= a, b, c, d < n. 
 * a, b, c, and d are distinct. 
 * nums[a] + nums[b] + nums[c] + nums[d] == target. You may return the answer in any order.
 * 
 * https://leetcode.com/problems/4sum/
 * T(c) -> O(n^4) , S(c) -> 2*n , for set and list data structure
 * Gave TLE
 */

    static List<List<Integer>> fourSum(int[] nums, int target) {
        int n = nums.length;
        
        // for storing unique arrays
        Set<List<Integer>> hset = new HashSet<>();
        
        for(int i=0;i<n;i++)
        {
            for(int j=i+1;j<n;j++)
            {
                for(int k=j+1;k<n;k++)
                {
                    for(int l=k+1;l<n;l++)
                    {
                        long sum = (long)nums[i] + nums[j];
                        sum += nums[k];
                        sum += nums[l];
                        if(sum == target)
                        {
                            List<Integer> temp = Arrays.asList(nums[i], nums[j], nums[k], nums[l]);
                            // before storing in set, sort array
                            Collections.sort(temp);
                            hset.add(temp);
                        }
                    }
                }
            }
        }
        
        // creating list from set data structure
        List<List<Integer>> result = new ArrayList<>(hset);
        return result;
    }

    /*
     * Using hashset for finding 4th element
     * T(c) -> O(N3*log(M)), where N = size of the array, M = no. of elements in the set.
     * S(c) -> O(2 * no. of the quadruplets)+O(N) , set data structure and a list to store the quads, set data structure to store the array elements.
     */
    static List<List<Integer>> fourSum(int[] nums, int target) {
        int n = nums.length;
        
        // for storing unique arrays
        Set<List<Integer>> hset = new HashSet<>();
        
        for(int i=0;i<n;i++)
        {
            for(int j=i+1;j<n;j++)
            {
                HashSet<Integer> tempSet = new HashSet<>();
                for(int k=j+1;k<n;k++)
                {
                    int fourth = target - (nums[i] + nums[j] + nums[k]);
                    if(tempSet.contains(fourth)) 
                    {
                        long sum = (long)nums[i] + nums[j];
                        sum += nums[k];
                        sum += fourth;
                        
                        if(sum == target)
                        {
                            List<Integer> tempList = Arrays.asList(nums[i], nums[j], nums[k], fourth);
                            // before storing in set, sort array
                            Collections.sort(tempList);
                            hset.add(tempList);
                        }
                    }
                    
                    // add in hashset
                    tempSet.add(nums[k]);
                }
            }
        }
        
        // creating list from set data structure
        List<List<Integer>> result = new ArrayList<>(hset);
        return result;
    }

    public static void main(String[] args) {
        int arr[] = {1,0,-1,0,-2,2};
        int target = 0;
        /* [[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]] */
        List<List<Integer>> ans = fourSum(arr, target);
        System.out.println(ans);
    }
    
}

File: Week3/Python/ArrayProblems/FourSum.py
"""
    Given an array nums of n integers, return an array of all the unique quadruplets [nums[a], nums[b], nums[c], nums[d]] 
    such that:
    0 <= a, b, c, d < n
    a, b, c, and d are distinct.
    nums[a] + nums[b] + nums[c] + nums[d] == target
    You may return the answer in any order.

    https://leetcode.com/problems/4sum/
"""


"""Giving TLE"""
def fourSum(nums, target):
    """ T(c) -> O(n^4), S(c) -> (2 * number of quadtruples) for set and array"""
    n = len(nums)
    result_set = set()

    for i in range(n):
        for j in range(i+1, n):
            for k in range(j+1, n):
                for l in range(k+1, n):
                    if nums[i] + nums[j] + nums[k] + nums[l] == target:
                        temp = [nums[i], nums[j], nums[k], nums[l]]
                        temp.sort()
                        result_set.add(tuple(temp))

    return [list(item) for item in result_set]


def fourSumBetter(nums, target):
    """ T(c) -> O(n^3), S(c) -> O(2 * number of quadtruples) for both a set and an array of quadruplets + O(n) for seen set element"""
    n = len(nums)
    # set for holding final result
    result_set = set()

    for i in range(n):
        for j in range(i+1, n):
            # set for finding the forth number
            seen = set()
            for k in range(j+1, n):
                another_num = target - (nums[i] + nums[j] + nums[k])
                if another_num in seen:
                    temp = [nums[i], nums[j], nums[k], another_num]
                    temp.sort()
                    result_set.add(tuple(temp))
                
                # add in the set
                seen.add(nums[k])

    return [list(item) for item in result_set]


def fourSumOptimal(nums, target):
    """ T(c) -> O(n^3), S(c) -> O(n^2) due to the storage of quadruplets in the result list """
    n = len(nums)
    
    # sort the array
    nums.sort()
    
    result = []
    
    # using two pointers approach
    for i in range(n):
        # skip duplicates of i
        if i > 0 and nums[i-1] == nums[i]:
            continue
            
        for j in range(i+1, n):
            # skip duplicates of j
            if j > i+1 and nums[j-1] == nums[j]:
                continue
                
            k = j+1
            l = n-1
            while k < l:
                total = nums[i] + nums[j] + nums[k] + nums[l]
                # total is less than target
                if total < target:
                    k += 1
                
                # total is greater than target
                elif total > target:
                    l -= 1
                
                # element found
                else:
                    temp = [nums[i], nums[j], nums[k], nums[l]]
                    result.append(temp)
                    k += 1
                    l -= 1
                    
                    # skip duplicates of k
                    while k < l and nums[k-1] == nums[k]:
                        k += 1
                    
                    # skip duplicates of l
                    while k < l and nums[l+1] == nums[l]:
                        l -= 1

    return result


arr = [1,0,-1,0,-2,2]
target = 0
print(arr, target)
print(fourSum(arr, target))
print(fourSumBetter(arr, target))
print(fourSumOptimal(arr, target))
